### PR TITLE
fix Issue 17568 - [scope] addresses to fields can be escaped from sco…

### PR DESCRIPTION
**Pull request by WalterBright**
…pe method

### CODE REVIEW

**Review by UplinkCoder — APPROVED** [review]
lgtm

File: test/fail_compilation/retscope2.d [hunk 1]
```diff
@@ -162,3 +162,25 @@ void foo800()
...
+
+int* p17568;
+struct T17568
+{
+    int a;
+    void escape() @safe scope
+    {
+        p17568 = &a;
+    }
+}
```
Comment (mathias-lang-sociomantic) [inline test/fail_compilation/retscope2.d]: Could you add a test case with a struct nested inside a function, and the pointer being at function scope ?

File: src/ddmd/escape.d [hunk 2]
```diff
@@ -173,7 +173,8 @@ bool checkParamArgumentEscape(Scope* sc, FuncDeclaration fdc, Identifier par, Ex
  */
 bool checkAssignEscape(Scope* sc, Expression e, bool gag)
 {
-    //printf("checkAssignEscape(e: %s)\n", e.toChars());
+    enum log = false;
+    if (log) printf("checkAssignEscape(e: %s)\n", e.toChars());
```
Comment (rainers) [inline src/ddmd/escape.d]: Better use `static if(log)` here?

File: test/fail_compilation/retscope2.d (hunk 1 above)
Comment (rainers) [inline test/fail_compilation/retscope2.d]: Here's a test as suggested by Mathias that still fails to fail to compile:
```
void test17568() @safe
{
	int* p;
	struct T
	{
		int a;
	}
	void escape(ref T t) @safe
	{
		p = &t.a;
	}
    {
        T t;
        escape(t);
    }
    auto bug = *p;
}
```

File: src/ddmd/escape.d [hunk 3]
```diff
@@ -299,15 +300,17 @@ bool checkAssignEscape(Scope* sc, Expression e, bool gag)
...
             continue;
 
         Dsymbol p = v.toParent2();
 
         // If va's lifetime encloses v's, then error
         if (va &&
-            (va.enclosesLifetimeOf(v) && !(v.storage_class & STCparameter) || va.storage_class & STCref) &&
+            (va.enclosesLifetimeOf(v) && !(v.storage_class & STCparameter) ||
+             va.storage_class & STCref ||
+             va.isDataseg()) &&
```
Comment (rainers) [inline src/ddmd/escape.d]: This looks like a special case very much tuned towards the test case. Shouldn't the check for global data be part of enclosesLifetimeOf?

File: src/ddmd/escape.d (hunk 2 above)
Comment (WalterBright) [inline src/ddmd/escape.d]: It doesn't make any difference there.

File: test/fail_compilation/retscope2.d (hunk 1 above)
Comment (WalterBright) [inline test/fail_compilation/retscope2.d]: It does compile, but should not. Filed as https://issues.dlang.org/show_bug.cgi?id=17725

File: src/ddmd/escape.d (hunk 3 above)
Comment (rainers) [inline src/ddmd/escape.d]: BTW: should all the changes to this condition also apply to the similar one at line 245?

Comment (MartinNowak) [inline src/ddmd/escape.d]: Indeed global data has infinite lifetime, thus an `if (isDataseg()) return true;` in `enclosesLifetimeOf` seems more appropriate.

**Review by MartinNowak — CHANGES_REQUESTED** [review]

File: src/ddmd/escape.d (hunk 3 above)
Comment (WalterBright) [inline src/ddmd/escape.d]: enclosesLifetimeOf presupposes an ordering that does not exist for statics. I.e. all statics have infinite lifetime, so there is no ordering. Locals, on the other hand, have an ordering.

Comment (WalterBright) [inline src/ddmd/escape.d]: Aside from that, this PR is meant to solve a specific issue. Redoing enclosesLifetimeOf would involve redoing all the rest of the logic. Something of that scope should be done as a separate refactoring.

Comment (WalterBright) [inline src/ddmd/escape.d]: > should all the changes to this condition also apply to the similar one at line 245?

The `va.isDataseg()` check is done a few lines down from that.

**Comment by mathias-lang-sociomantic** [comment]
Ping @WalterBright (Martin's feedback)

BTW I created a `#scope` channel on slack if you ever want to join us :)

**Comment by WalterBright** [comment]
I replied to @MartinNowak 's feedback. Please take a look at my comments.

**Comment by ZombineDev** [comment]
Ping @MartinNowak @rainers can we move forward with this?

**Comment by rainers** [comment]
> enclosesLifetimeOf presupposes an ordering that does not exist for statics

enclosesLifetimeOf pretends to do more than it does (by name and documentation). Maybe it should just be a local function inside checkAssignExp.

Currently, you shouldn't be calling enclosesLifetimeOf on statics at all, so check isDataSeg() before would be cleaner. Otherwise it returns random results.

I'm ok with merging, but maybe the bugzilla title should be narrowed down to this case (for the changelog), https://issues.dlang.org/show_bug.cgi?id=17725 is another example of the same issue. Not sure, if there are more.

**Comment by WalterBright** [comment]
It works correctly regardless of the order, because those functions are pure and they're OR'd together.

You can change the title if you like. But I want to get this pulled, so I can work on the next [scope] issue. I tried working on several in parallel, but it wound up being a lot of confusing rebase extra work.

Because there are more scope issues in bugzilla, I don't care if any of these fixes are not models of clarity. I want to get it working, and have a test suite to cover the cases. Afterwards is the right time to take a step back and refactor it. Constantly refactoring before knowing what the end result needs to look like is a lot of churn.

Taking a month to pull one fix that is **one line** of code, when there are many in the queue, is a disaster. I have other PRs that consist merely of adding comments that sit around for weeks at a time, too.

**Comment by MartinNowak** [comment]
> You can change the title if you like. But I want to get this pulled, so I can work on the next [scope] issue. I tried working on several in parallel, but it wound up being a lot of confusing rebase extra work.
> 
> Because there are more scope issues in bugzilla, I don't care if any of these fixes are not models of clarity. I want to get it working, and have a test suite to cover the cases. Afterwards is the right time to take a step back and refactor it. Constantly refactoring before knowing what the end result needs to look like is a lot of churn.
> 
> Taking a month to pull one fix that is one line of code, when there are many in the queue, is a disaster. I have other PRs that consist merely of adding comments that sit around for weeks at a time, too.

IMHO an indication that you shouldn't submit each commit as individual PR ;), but I don't want to open that discussion again.

We've just asked for the `isDataseg()` check to be moved into `enclosesLifetimeOf`, as any static variable has infinite lifetime and thus outlives anything else, it's the infimum of the partially ordered sequenceNumbers.
It smells like there could be more manifestation of the missing `isDataseg()` check, hence the request, which seems both trivial and reasonable.
https://github.com/dlang/dmd/blob/4d86fcba2fd2ef86cc85738cd2ac2b059dbb5800/src/ddmd/declaration.d#L2450
If you have a more strict understanding of the term `encloses`, that's fine, and indeed rewriting all the exisiting code that checks `enclosesLifetimeOf` and `isDataSeg` separately is out of scope here.

**Review by MartinNowak — APPROVED** [review]

**Pull request merged**